jlm/rvsdg/bitstring/arithmetic.hpp

```diff
@@ -12,91 +12,193 @@
-    create(size_t nbits) const override;                                                       \
-                                                                                               \
-    static inline output *                                                                     \
-    create(size_t nbits, output * op1, output * op2)                                           \
-    {                                                                                          \
-      return simple_node::create_normalized(op1->region(), NAME##_op(nbits), { op1, op2 })[0]; \
-    }                                                                                          \
-  };
-
-DECLARE_BITUNARY_OPERATION(bitneg)
-DECLARE_BITUNARY_OPERATION(bitnot)
-
-DECLARE_BITBINARY_OPERATION(bitadd)
-DECLARE_BITBINARY_OPERATION(bitand)
-DECLARE_BITBINARY_OPERATION(bitashr)
-DECLARE_BITBINARY_OPERATION(bitmul)
-DECLARE_BITBINARY_OPERATION(bitor)
-DECLARE_BITBINARY_OPERATION(bitsdiv)
-DECLARE_BITBINARY_OPERATION(bitshl)
-DECLARE_BITBINARY_OPERATION(bitshr)
-DECLARE_BITBINARY_OPERATION(bitsmod)
-DECLARE_BITBINARY_OPERATION(bitsmulh)
-DECLARE_BITBINARY_OPERATION(bitsub)
-DECLARE_BITBINARY_OPERATION(bitudiv)
-DECLARE_BITBINARY_OPERATION(bitumod)
-DECLARE_BITBINARY_OPERATION(bitumulh)
-DECLARE_BITBINARY_OPERATION(bitxor)
+template<typename reduction, const char * name>
+class make_bitunary_op final : public bitunary_op
+{
+public:
+  ~make_bitunary_op() noexcept override;
+
+  inline make_bitunary_op(const bittype & type) noexcept
+      : bitunary_op(type)
+  {}
+
+  virtual bool
+  operator==(const operation & other) const noexcept override;
+
+  virtual bitvalue_repr
+  reduce_constant(const bitvalue_repr & arg) const override;
+
+  virtual std::string
+  debug_string() const override;
+
+  virtual std::unique_ptr<operation>
+  copy() const override;
+
+  virtual std::unique_ptr<bitunary_op>
+  create(size_t nbits) const override;
+
+  static inline output *
+  create(size_t nbits, output * op)
+  {
+    return simple_node::create_normalized(op->region(), make_bitunary_op(nbits), { op })[0];
+  }
+};
+
+template<typename reduction, const char * name, enum binary_op::flags opflags>
+class make_bitbinary_op final : public bitbinary_op
+{
+public:
+  ~make_bitbinary_op() noexcept override;
+
+  inline make_bitbinary_op(const bittype & type) noexcept
+      : bitbinary_op(type)
+  {}
+
+  virtual bool
+  operator==(const operation & other) const noexcept override;
+
+  virtual enum binary_op::flags
+  flags() const noexcept override;
+
+  virtual bitvalue_repr
+  reduce_constants(const bitvalue_repr & arg1, const bitvalue_repr & arg2) const override;
+
+  virtual std::string
+  debug_string() const override;
+
+  virtual std::unique_ptr<operation>
+  copy() const override;
+
+  virtual std::unique_ptr<bitbinary_op>
+  create(size_t nbits) const override;
+
+  static inline output *
+  create(size_t nbits, output * op1, output * op2)
```

Comment:
Remove `inline`

Reply:
I know that inline is implicit when stated in class declaration, although I generally prefer to keep it to make it clear that I did not put it into the class decl by accident but by intent of making it inline. Removed it for now anyways.